src/RustyObjectStore.jl

```diff
@@ -1406,6 +1417,285 @@ function _unsafe_write(stream::WriteStream, input::Ptr{UInt8}, nbytes::Int; flus
+It is necessary to `finish!` the stream if it is not run to completion.
+"""
+mutable struct ListStream
+    ptr::Ptr{Nothing}
+    ended::Bool
+    error::Option{String}
+end
+
+function stream_end(stream::ListStream)
+    @assert (!stream.ended && isnothing(stream.error))
+    stream.ended = true
+    @ccall rust_lib.destroy_list_stream(stream.ptr::Ptr{Nothing})::Cint
+end
+
+function stream_error(stream::ListStream, err::String)
+    @assert (!stream.ended && isnothing(stream.error))
+    stream.error = err
+    @ccall rust_lib.destroy_list_stream(stream.ptr::Ptr{Nothing})::Cint
+end
+
+"""
+    list_objects_stream(prefix, conf) -> ListStream
+Send a list request to the object store returning a stream of entry chunks.
+# Arguments
+- `prefix::String`: Only objects with this prefix will be returned.
+- `conf::AbstractConfig`: The configuration to use for the request.
+  It includes credentials and other client options.
+# Keyword Arguments
+- `offset::Option{String}`: (Optional) Start listing after this offset
+# Returns
+- `stream::ListStream`: The stream of object metadata chunks.
+# Throws
+- `ListException`: If the request fails for any reason.
+"""
+function list_objects_stream(prefix::String, conf::AbstractConfig; offset::Option{String} = nothing)
+    response = ListStreamResponseFFI()
+    ct = current_task()
+    event = Base.Event()
+    handle = pointer_from_objref(event)
+    config = into_config(conf)
+    c_offset = if isnothing(offset)
+        C_NULL
+    else
+        offset
+    end
+    while true
+        preserve_task(ct)
+        result = GC.@preserve config response event try
+            result = @ccall rust_lib.list_stream(
+                prefix::Cstring,
+                c_offset::Cstring,
+                config::Ref{Config},
+                response::Ref{ListStreamResponseFFI},
+                handle::Ptr{Cvoid}
+            )::Cint
+
+            wait_or_cancel(event, response)
+
+            result
+        finally
+            unpreserve_task(ct)
+        end
+
+        if result == 2
+            # backoff
+            sleep(0.01)
+            continue
+        end
+
+        # No need to destroy_list_stream in case of errors here
+        @throw_on_error(response, "list_stream", ListException)
+
+        return ListStream(response.stream, false, nothing)
+    end
+end
+
+"""
+    next_chunk!(stream) -> Option{Vector{ListEntry}}
+Fetch the next chunk from a ListStream.
+An empty chunk indicates end of stream.
+After an error any following calls will replay the error.
+# Arguments
+- `stream::ListStream`: The stream of object metadata list chunks.
+# Returns
+- `entries::Vector{ListEntry}`: The array with metadata for each object in the prefix.
+  Resturns and empty array if no objects match or the stream is over.
+# Throws
+- `ListException`: If the request fails for any reason.
+"""
+function next_chunk!(stream::ListStream)
```

Comment:
TODO: mark stream at ended if it returns > 0 entries but we know they are the last ones.